proposals/NNNN-swiftpm-package-traits.md

```diff
@@ -0,0 +1,479 @@
+        url: String,
+        exact version: Version,
+        traits: Traits
+    ) -> Package.Dependency
+
+    // MARK: Registry 
+
+    public static func package(
+        id: String,
+        from version: Version,
+        traits: Traits
+    ) -> Package.Dependency
+
+    public static func package(
+        id: String,
+        exact version: Version,
+        traits: Traits
+    ) -> Package.Dependency
+
+    public static func package(
+        id: String,
+        _ range: Range<Version>,
+        traits: Traits
+    ) -> Package.Dependency
+
+    public static func package(
+        id: String,
+        _ range: ClosedRange<Version>,
+        traits: Traits
+    ) -> Package.Dependency
+}
+```
+
+### Default traits
+
+Default traits allow package authors to define a set of traits that they think
+cater to the majority use-cases of the package. When choosing the initial
+default traits or adding a new default trait it is important to consider that
+removing a default trait is a SemVer major change since it can potentially
+remove APIs.
+
+### Trait specific command line options for `swift build/run`
+
+When executing one of `swift build/run` options can be passed to control
+which traits for the root package are enabled:
+
+- `--traits` _TRAITS_: Enables the passed traits of the package. Multiple traits can be
+  specified by providing a comma separated list e.g. `--traits Trait1,Trait2`.
+- `--enable-all-traits`: Enables all traits of the package.
+- `--disable-default-traits`: Disables all default traits of the package.
+
+
+### Trait limitations
+
+To prevent abuse, limit the complexity and make sure it integrates with the
+compiler a few limitations are imposed.
+
+#### Number of traits
+
+[Other
+ecosystems](https://blog.rust-lang.org/2023/10/26/broken-badges-and-23k-keywords.html)
+have shown that a large number of traits can have significant impact on
+registries and dependency managers. To avoid such a scenario an initial maximum
+number of traits of 300 per package is imposed. This can be revisited later once
+traits have been used in the ecosystem extensively.
+
+### Allowed characters for trait names
+
+Since traits can show up both in the `Package.swift` and in source code when
+checking if a trait is enabled, the allowed characters for a trait name have to
+be restricted. The following rules are enforced on trait names:
+
+- The first character must be a [Unicode XID start
+  character](https://unicode.org/reports/tr31/#Figure_Code_Point_Categories_for_Identifier_Parsing)
+  (most letters), a digit, or `_`.
+- Subsequent characters must be a [Unicode XID continue
+  character](https://unicode.org/reports/tr31/#Figure_Code_Point_Categories_for_Identifier_Parsing)
+  (a digit, `_`, or most letters), `-`, or `+`.
+
+## Impact on existing packages
+
+There is no impact on existing packages. Any package can start adopting package
+traits but in doing so **must not** move existing API behind new traits.
```

Comment:
Even if the trait is a default trait?

Reply:
Yes, it is never safe. An adopter might declare `disableDefaultTraits: true`; hence, loosing the API that got moved behind the default.

Reply:
Wouldn't that be the adopter breaking the API rather than the package?